libsolidity/ExpressionCompiler.cpp

```diff
@@ -628,7 +628,33 @@ void ExpressionCompiler::appendTypeConversion(Type const& _typeOnStack, Type con
 		return;
 	Type::Category stackTypeCategory = _typeOnStack.getCategory();
 	Type::Category targetTypeCategory = _targetType.getCategory();
-	if (stackTypeCategory == Type::Category::INTEGER || stackTypeCategory == Type::Category::CONTRACT ||
+
+	if (stackTypeCategory == Type::Category::STRING && targetTypeCategory == Type::Category::INTEGER)
+	{
+		// conversion from string to hash. no need to clean the high bit
+		// only to shift right because of opposite alignment
+		IntegerType const& targetIntegerType = dynamic_cast<IntegerType const&>(_targetType);
+		StaticStringType const& sourceStringType = dynamic_cast<StaticStringType const&>(_typeOnStack);
+		if (targetIntegerType.isHash())
+		{
+			solAssert(targetIntegerType.getNumBits() == sourceStringType.getNumBytes() * 8, "The size should be the same.");
+			m_context << u256(std::pow(2, 256 - sourceStringType.getNumBytes() * 8)) <<
```

Comment:
I fear this might overflow for e.g. string1 to hash8 conversion. Please add a test for this case and change this to
something like `u256(1) << (256 - sourceStringType.getNumBytes() * 8)`.